bundles/org.openhab.binding.cbus/src/main/java/org/openhab/binding/cbus/internal/discovery/CBusGroupDiscovery.java

```diff
@@ -0,0 +1,105 @@
+                Map<Integer, ThingTypeUID> applications = new HashMap<Integer, ThingTypeUID>();
+                applications.put(CBusBindingConstants.CBUS_APPLICATION_LIGHTING, CBusBindingConstants.THING_TYPE_LIGHT);
+                applications.put(CBusBindingConstants.CBUS_APPLICATION_DALI, CBusBindingConstants.THING_TYPE_DALI);
+                applications.put(CBusBindingConstants.CBUS_APPLICATION_TEMPERATURE,
+                        CBusBindingConstants.THING_TYPE_TEMPERATURE);
+                applications.put(CBusBindingConstants.CBUS_APPLICATION_TRIGGER,
+                        CBusBindingConstants.THING_TYPE_TRIGGER);
+
+                Network network = cbusNetworkHandler.getNetwork();
+                if (network == null) {
+                    return;
+                }
+                for (Map.Entry<Integer, ThingTypeUID> applicationItem : applications.entrySet()) {
+                    Application application = network.getApplication(applicationItem.getKey());
+                    if (application == null) {
+                        continue;
+                    }
+                    ArrayList<Group> groups = application.getGroups(false);
+                    for (Group group : groups) {
+                        logger.debug("Found group: {} {} {}", application.getName(), group.getGroupID(),
+                                group.getName());
+                        Map<String, Object> properties = new HashMap<>();
+                        properties.put(CBusBindingConstants.PROPERTY_APPLICATION_ID,
+                                Integer.toString(applicationItem.getKey()));
+                        properties.put(CBusBindingConstants.PROPERTY_GROUP_ID, Integer.toString(group.getGroupID()));
+                        properties.put(CBusBindingConstants.PROPERTY_GROUP_NAME, group.getName());
+                        properties.put(CBusBindingConstants.PROPERTY_NETWORK_ID,
+                                Integer.toString(network.getNetworkID()));
+
+                        ThingUID bridgeUid = cbusNetworkHandler.getThing().getBridgeUID();
+                        if (bridgeUid != null) {
+                            ThingUID uid = new ThingUID(applicationItem.getValue(),
+                                    Integer.toString(group.getGroupID()), bridgeUid.getId(),
+                                    cbusNetworkHandler.getThing().getUID().getId());
+                            DiscoveryResult result = DiscoveryResultBuilder.create(uid).withProperties(properties)
+                                    .withLabel("CBUS " + group.getName() + "(" + group.getGroupID() + ")")
+                                    .withBridge(cbusNetworkHandler.getThing().getUID()).build();
+                            thingDiscovered(result);
+                        }
+                    }
+                }
+            } catch (CGateException e) {
+                logger.warn("Failed to discover groups", e);
+            }
+        }
+    }
+
```

Comment:
You should override `stopScan` as well.

Reply:
The code in startscan will finish in a finite time or fail with a communication error that will throw an exception long before that timeout value. So i have set the timeout to 0 and in that case i dont believe i need to implement stopScan.

Reply:
how long is a 'finite' time? It would be against user expectations for discovery results appear after they thought they stopped scanning.

Reply:
In my case less than a second. I cant see it ever taking more than 10 or so. Certainly significantly less than the 300 second timeout that was there.

Reply:
Just to clarify my understanding of this. If i do a scan from paperUi it seems to call StopScan presumably to stop any scans that might be running. If i set a timeout then stopscan will be called after that timeout but there is no well defined time for that timeout. The spinner in the Paper UI search page runs for a period of time which is unrelated to the timeout and when it stop it doesnt call stopscan so results may continue to come in depending on what timeout value was chosen by the binding? The results can also come in after someone leaves the page which is the only way they really would believe they stopped scanning since again it wont call stopScan and the timeout chosen seems arbitrary. Have i missed something here? I could implement something to look at a shared variable to stop scanning but given how arbitrary the whole mechanism is it seems like a lot of work for something i dont think will be triggered here ever other than people pressing the search button more than once very quickly.

Reply:
Hmm, I looked into it and it seems that the only way you would be able to indicate that a scan is finished is to call `stopScan` yourself. But since that doesn't look like it would cancel the scheduled called to `stopScan` you would need to call another method to stop that task.

I recommend you add this method and then call it when you are finished scanning.
```
private synchronized void scanFinished(){
    stopScan();//this notifies the scan listener that the scan is finished
    abortScan();//this clears the scheduled call to stopScan
}
```

I don't know if the UI will react appropriately but this way would the appropriate way for the code to handle it. If the UI still doesn't react appropriately then I would blame the UI not the code at that point.

Reply:
I'm not too familiar with how the UI interacts with the codebase so I can't give you good answers with regards to when it calls what.
What I can tell you is that the UI calls `startScan(ScanListener)` which at that point the AbstractDiscoveryService itself sets it's own timer for when to call `stopScan`. The timeout is specified by subclasses when calling the AbstractDiscoveryService constructor.

Reply:
Implemented stopscan and scanFinished as suggested. Tested by adding sleeps into the code to delay things and it seems to work (ignoring the UI issues).